server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -206,18 +206,58 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
-        final ActionListener<BulkResponse> releasingListener = ActionListener.runBefore(listener, releasable::close);
+        final boolean isOnlySystem = isOnlySystem(bulkRequest, initialState.metadata().getIndicesLookup(), systemIndices);
         final String executorName = isOnlySystem ? Names.SYSTEM_WRITE : Names.WRITE;
-        threadPool.executor(Names.WRITE).execute(new ActionRunnable<>(releasingListener) {
-            @Override
-            protected void doRun() {
-                doInternalExecute(task, bulkRequest, executorName, releasingListener);
+        final Releasable releasable = indexingPressure.markCoordinatingOperationStarted(indexingOps, indexingBytes, isOnlySystem);
+        // We should use the releasingListener from here onwards
+        final ActionListener<BulkResponse> releasingListener = ActionListener.runBefore(originalListener, releasable::close);
+
+        final ClusterBlockException blockException = initialState.blocks().globalBlockedException(ClusterBlockLevel.WRITE);
+        if (blockException != null) {
+            if (false == blockException.retryable()) {
+                releasingListener.onFailure(blockException);
+                return;
             }
-        });
+            logger.trace("cluster is blocked, waiting for it to recover", blockException);
+            final ClusterStateObserver clusterStateObserver = new ClusterStateObserver(
+                initialState,
+                clusterService,
+                bulkRequest.timeout(),
+                logger,
+                threadPool.getThreadContext()
+            );
+            clusterStateObserver.waitForNextChange(new ClusterStateObserver.Listener() {
+                @Override
+                public void onNewClusterState(ClusterState state) {
+                    threadPool.executor(Names.WRITE).execute(new ActionRunnable<>(releasingListener) {
+                        @Override
+                        protected void doRun() {
+                            doInternalExecute(task, bulkRequest, executorName, releasingListener);
+                        }
+                    });
+                }
+
+                @Override
+                public void onClusterServiceClose() {
+                    releasingListener.onFailure(new NodeClosedException(clusterService.localNode()));
+                }
+
+                @Override
+                public void onTimeout(TimeValue timeout) {
+                    releasingListener.onFailure(blockException);
+                }
+            }, newState -> false == newState.blocks().hasGlobalBlockWithLevel(ClusterBlockLevel.WRITE));
+        } else {
+            threadPool.executor(Names.WRITE).execute(new ActionRunnable<>(releasingListener) {
+                @Override
+                protected void doRun() {
+                    doInternalExecute(task, bulkRequest, executorName, releasingListener);
+                }
+            });
```

Comment:
This is the same as the `onNewClusterState` implementation, can we find a way to reduce the duplication?

Also I wonder why we're forking to `WRITE` here even though we know it might be a low-latency system-indices-only write. Not a new problem, but something to think about in future.

Reply:
It is now splitted into a separate method. I didn't do that because it felt too small for a method and not too big for a duplication. It is also possible to wrap the thing in a runnable so that it can be called in both places. But it feels wasteful to create an extra object. Please let me know if you have a preference.

 > Also I wonder why we're forking to WRITE here even though we know it might be a low-latency system-indices-only write. Not a new problem, but something to think about in future.

Yeah. I noticed it as well when I duplicated the code. I think it is a bug. But would rather address it separately.